### PR TITLE
fix: make sure data is json serializable

### PR DESCRIPTION
**Pull request by ogabrielluiz**
This pull request includes two commits. The first commit enhances the data serialization process by adding recursive handling in the `to_json` method. This allows for better handling of `datetime` objects by converting them to ISO format. The second commit refactors the `recursive_serialize_or_str` function into a separate module called `serialize.py`. This improves modularity and reusability of the code. The function is moved from `schema.py` to the new module, and the imports in `data.py`, `artifact.py`, and `schema.py` are updated accordingly.

### CODE REVIEW

**Comment by edwinjosechittilappilly** [comment]
Screenshot 2024-10-24 at 12 13 37 PM

Tested this PR with the error that I had in PR #4258 (By merging both changes in a new branch)
Still causes the issue.

**Comment by edwinjosechittilappilly** [comment]
> Screenshot 2024-10-24 at 12 13 37 PM
> Tested this PR with the error that I had in PR #4258 (By merging both changes in a new branch) Still causes the issue.

made changes in branch #4258 Those errors were from the function change in _process_connection_or_parameters

Ignore the above comment.

**Review by edwinjosechittilappilly — APPROVED** [review]
LGTM

File: src/backend/base/langflow/schema/serialize.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+            return {k: recursive_serialize_or_str(v) for k, v in obj.dict().items()}
+        if hasattr(obj, "model_dump"):
+            return {k: recursive_serialize_or_str(v) for k, v in obj.model_dump().items()}
+        if isinstance(obj, type) and issubclass(obj, BaseModel):
+            # This a type BaseModel and not an instance of it
+            return repr(obj)
+        return str(obj)
+    except Exception:  # noqa: BLE001
+        logger.debug(f"Cannot serialize object {obj}")
+        return str(obj)
```
Comment (jordanrfrazier) [inline src/backend/base/langflow/schema/serialize.py:43]: Curious the downstream effects of this. Is this currently used for serializing an object for caching?

**Pull request merged**